[Merged by Bors] - refactor(SimpleGraph): make `completeGraph` and `emptyGraph` `abbrev`s for `⊤` and `⊥`

FordUniver opened the pull request:
define Top and Bot in-line for SimpleGraph and set completeGraph and emptyGraph as abbrev

---
Strangely `DecidableRel (completeGraph V).Adj` does not seem to be automatically inferred from `Top.adjDecidable` even though `⊤ : SimpleGraph V` is defined as `completeGraph V`. Am I missing something here or is it as simple as adding these two lines to mathlib?

YaelDillies commented:
I'll leave it to Kyle to decide whether having `completeGraph` is worth if we need to duplicate all the API we already have on `⊤`

FordUniver commented:
I'd happy with an approach that doesn't require duplicate APIs in mathlib (or in a particular project's code).

YaelDillies commented:
For your particular project, you can simply always use `⊤` instead of `completeGraph`

FordUniver commented:
> For your particular project, you can simply always use `⊤` instead of `completeGraph`

Yeah, that's what we're doing currently. But it feels... wrong. I thought for a minute `⊤` was actually its own definition and `completeGraph` should actually just point to that rather than giving the blackboard definition. But it's `⊤` that points to `completeGraph`, so it feels off and not "mathematician-friendly" to then use `⊤` in code.

YaelDillies commented:
> I thought for a minute `⊤` was actually its own definition and `completeGraph` should actually just point to that rather than giving the blackboard definition. But it's `⊤` that points to `completeGraph`, so it feels off and not "mathematician-friendly" to then use `⊤` in code.

But it is indeed `completeGraph` that points to `⊤`, not the other way around. cf the docstring of `completeGraph`:
> The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent. In Mathlib, this is usually referred to as `⊤`.

FordUniver commented:
> But it is indeed `completeGraph` that points to `⊤`, not the other way around. cf the docstring of `completeGraph`:

I actually never understood that docstring. In code `completeGraph` is straight up defined as a `SimpleGraph` by pretty much giving the blackboard definition. Only then is `CompleteAtomicBooleanAlgebra` established by providing `completeGraph` as `top`, again pretty much how you would do it on a blackboard. To then go and always invoke `⊤`, even when the statement has nothing do with `CompleteAtomicBooleanAlgebra` seems like taking one extra step? So actually `Top.adjDecidable` should be derived from `ompleteGraph.adjDecidable` so one can use it without invoking `CompleteAtomicBooleanAlgebra`.

YaelDillies commented:
> > But it is indeed `completeGraph` that points to `⊤`, not the other way around. cf the docstring of `completeGraph`:
> 
> I actually never understood that docstring. In code `completeGraph` is straight up defined as a `SimpleGraph` by pretty much giving the blackboard definition. Only then is `CompleteAtomicBooleanAlgebra` established by providing `completeGraph` as `top`, again pretty much how you would do it on a blackboard. To then go and always invoke `⊤`, even when the statement has nothing do with `CompleteAtomicBooleanAlgebra` seems like taking one extra step? So actually `Top.adjDecidable` should be derived from `ompleteGraph.adjDecidable` so one can use it without invoking `CompleteAtomicBooleanAlgebra`.

How it is defined is definitely an implementation, especially if it conflicts with the docstring

urkud commented:
@YaelDillies I think, the suggestion here is to define `Top` instance inline, then define `completeGraph` as an `abbrev` for `\top`.

YaelDillies commented:
Sure, that's a good suggestion to reduce API surface. @FordUniver, would you be willing to take that on? This way, you don't need to add any decidability instance.

FordUniver commented:
Done, also replaced all `⊤ : SimpleGraph` and `⊥ : SimpleGraph` occurrences in mathlib and modified the docstring. One thing I was unsure about is why these two lines where there:

```
attribute [aesop safe (rule_sets := [SimpleGraph])] Ne.symm
attribute [aesop safe (rule_sets := [SimpleGraph])] Ne.irrefl
```

I removed them and nothing broke 🤷‍♂️

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Basic.lean:
```diff
@@ -134,16 +134,6 @@ theorem SimpleGraph.fromRel_adj {V : Type u} (r : V → V → Prop) (v w : V) :
     (SimpleGraph.fromRel r).Adj v w ↔ v ≠ w ∧ (r v w ∨ r w v) :=
   Iff.rfl
 
-attribute [aesop safe (rule_sets := [SimpleGraph])] Ne.symm
-attribute [aesop safe (rule_sets := [SimpleGraph])] Ne.irrefl
```
Sorry, just saw https://github.com/leanprover-community/mathlib4/pull/23838#issuecomment-2853800630 now. You will need to ask @kmill, but regardless it's out of scope for the current PR 😄

FordUniver replied:
Fair, they're back now

YaelDillies replied:
Can you move them back to where they were?

FordUniver replied:
Done

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -590,7 +590,7 @@ def topEquiv : (⊤ : Subgraph G).coe ≃g G where
 /-- The bottom of the `Subgraph G` lattice is equivalent to the empty graph on the empty
 vertex type. -/
-def botEquiv : (⊥ : Subgraph G).coe ≃g (⊥ : SimpleGraph Empty) where
+def botEquiv : (⊥ : Subgraph G).coe ≃g (emptyGraph Empty) where
```
```suggestion
def botEquiv : (⊥ : Subgraph G).coe ≃g emptyGraph Empty where
```

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge

bryangingechen reviewed (approved):
Thanks!
bors r+
